modules/storages/app/common/storages/peripherals/storage_interaction/authentication.rb

```diff
@@ -0,0 +1,69 @@
+# frozen_string_literal:true
+
+#-- copyright
+# OpenProject is an open source project management software.
+# Copyright (C) 2012-2024 the OpenProject GmbH
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License version 3.
+#
+# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
+# Copyright (C) 2006-2013 Jean-Philippe Lang
+# Copyright (C) 2010-2013 the ChiliProject Team
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+module Storages
+  module Peripherals
+    module StorageInteraction
+      class Authentication
+        using ::Storages::Peripherals::ServiceResultRefinements
+
+        def self.[](strategy)
+          case strategy.key
+          when :basic_auth
+            AuthenticationStrategies::BasicAuth.new
+          when :oauth_user_token
+            AuthenticationStrategies::OAuthUserToken.new(strategy.user)
+          when :oauth_client_credentials
+            AuthenticationStrategies::OAuthClientCredentials.new
+          else
+            raise "Invalid authentication strategy '#{strategy}'"
+          end
+        end
+
+        # Checks for the current authorization state of a user on a specific file storage.
+        # Returns one of three results:
+        # - :connected If a valid user token is available
+        # - :failed_authorization If a user token is available, which is invalid and not refreshable
+        # - :error If an unexpected error occurred
+        def self.authorization_state(storage:, user:)
```

Comment:
ignore this method, I simply didn't want to cut this out into a separate commit. It is needed later, when authorization state is extracted from the connection manager